### PR TITLE
Fixes #5994 - QueuedThreadPool "free" threads

### LINKED ISSUE #5994: QueuedThreadPool "free" threads
**Jetty version**
9.4.x

**Description**
On large machines, the heuristics we have for number of acceptors, selectors, and reserved threads may "steal" a lot of threads from the thread pool, so that a metric such as `(threads - idleThreads)/maxThreads` may show large percentages even for a completely idle server.

For example, for a 112 cores machine we have 
* 4 acceptors
* 56 selectors
* 112 reserved

for a total of 172 "stolen" threads.
With `maxThreads=1024` or similar, that is a 17% "utilization" even if the server is idle.

Using `busyThreads` instead of `(threads - idleThreads)` takes into account reserved threads, so that figure would be only 60 "stolen" threads for a 6% "utilization".

I wonder if we should at exclude acceptors and selectors from the count (i.e. rather than taking them from the pool, allocate the threads outside the pool -- however this would make the thread budget component irrelevant now).

Excluding them would help people to answer the question: "should I increase `maxThreads` or I'm good?".
Users that monitor `busyThreads/maxThreads` will know on a 0-100% scale rather than always having a 6-17% "noise" even when the server is idle.

@gregw @joakime @lorban thoughts?

### PR DESCRIPTION
**Pull request by sbordet**
Introduced to QueuedThreadPool:

* getMaxReservedThreads()
* getAvailableReservedThreads()
* getAvailableThreads()
* getReadyThreads()
* getLeasedThreads()

Also few small code cleanups.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

### CODE REVIEW

**Review by gregw — CHANGES_REQUESTED** [review]
Updated from my comment on the issue...

We need a good vocab to describe all the states a thread can be in and then matching getters and JMX fields:

* inactive: not started but there is capacity to start another thread
* idle: started and waiting in the pool for a task
* reserved: taken from the pool but reserved to do a task
* leased: taken from the pool to do a (budgeted - maxReserved) task like selecting
* utilised: taken from the pool for a non-leased task.

So we have some invariants and derived stats:
```
 max = inactive + idle + reserved + leased + utilised
 threads = idle + reserved + leased + utilised
 busy = leased + utilised
 ready = idle + reserved
 utilised = threads - idle - reserved - leased 
 available = max - busy
 maxAvailable = max - leased
 utilisation = utilised / maxAvailable # 0.0 means idle, 1.0 means at capacity 
```

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java [hunk 1]
```diff
@@ -425,16 +424,34 @@ public int getMinThreads()
      * @return number of reserved threads or or -1 for heuristically determined
      * @see #setReservedThreads
      */
-    @ManagedAttribute("the number of reserved threads in the pool")
+    @ManagedAttribute("number of configured reserved threads")
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java]: Should we deprecate this?  It really is the same as maxReserved threads.  Or should we just change it to be the available reserved threads.

File: jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java [hunk 2]
```diff
@@ -425,16 +424,34 @@ public int getMinThreads()
...
-                return reservedThreadExecutor.getCapacity();
+            ReservedThreadExecutor reservedThreadExecutor = (ReservedThreadExecutor)tryExecutor;
+            return reservedThreadExecutor.getCapacity();
         }
-        return _reservedThreads;
+        return 0;
+    }
+
+    @ManagedAttribute("number of available reserved threads")
+    public int getAvailableReservedThreads()
```
Comment (gregw) [inline jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:524]: I don't like this name.  There is no such thing as an unavailable reserved thread.

**Review by gregw — CHANGES_REQUESTED** [review]

**Comment by joakime** [comment]
I'm keen to +1 this PR just because it's javadoc'd! :-)

**Review by gregw — APPROVED** [review]
I think there is more we can say in the javadoc.... but for another PR.

**Pull request merged**